File: app/src/main/java/io/duckduckgosearch/app/SearchActivity.java
package io.duckduckgosearch.app;

import android.app.Activity;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.webkit.WebViewClient;
import android.widget.AutoCompleteTextView;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.FragmentManager;
import androidx.room.Room;

public class SearchActivity extends AppCompatActivity implements WebViewFragment.OnSearchTermChange,
        AutoCompleteAdapter.OnItemClickListener, WebViewFragment.OnWebViewError, ErrorFragment.OnReloadButtonClick,
        WebViewFragment.OnPageFinish {

    AutoCompleteTextView searchBar;
    FragmentManager fragmentManager;
    ProgressBar progressBar;
    Activity activity;
    ImageView duckLogo;
    ImageButton eraseTextButton;
    RelativeLayout searchBarRoot;
    String latestTerm = "";
    String intentSearchTerm;
    WebViewFragment webViewFragment;
    AutoCompleteAdapter adapter;
    InputMethodManager manager;
    HistoryDatabase historyDatabase;
    boolean darkTheme;
    boolean fromIntent;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (PrefManager.isDarkTheme(this)) {
            setTheme(R.style.AppTheme_Dark_NoActionBar);
            darkTheme = true;
        }
        setContentView(R.layout.activity_search);

        historyDatabase = Room.databaseBuilder(this, HistoryDatabase.class, HistoryFragment.HISTORY_DB_NAME)
                .build();

        manager = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);

        duckLogo = findViewById(R.id.duck_logo);

        fragmentManager = getSupportFragmentManager();

        activity = this;

        searchBarRoot = findViewById(R.id.search_bar_edittext_root);

        progressBar = findViewById(R.id.search_progress);

        searchBar = findViewById(R.id.search_bar_edittext);

        adapterUpdate();

        eraseTextButton = findViewById(R.id.erase_button);

        Bundle bundle = getIntent().getExtras();
        if (bundle != null) {
            if (bundle.containsKey("search_term")) {
                intentSearchTerm = bundle.getString("search_term");
                fromIntent = true;
                search(intentSearchTerm);
            }
        }

        if (darkTheme) {
            findViewById(R.id.frame_layout).setBackgroundColor(
                    getResources().getColor(R.color.darkThemeColorPrimary));
            eraseTextButton.setImageDrawable(getResources().getDrawable(R.drawable.ic_outline_close_24px_white));
        }

        if (savedInstanceState == null && !fromIntent) {
            searchBar.requestFocus();
        } else {
            getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_HIDDEN);
            if (!searchBar.hasFocus()) {
                duckLogo.setVisibility(View.VISIBLE);
            }
        }

        searchBar.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                if (!v.getText().toString().equals("") && actionId == EditorInfo.IME_ACTION_DONE) {
                    search(v.getText().toString());
                    if (adapter != null) {
                        adapter.notifyDataSetChanged();
                    }
                    return true;
                }
                return false;
            }
        });
        searchBar.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (hasFocus) {
                    duckLogo.setVisibility(View.GONE);
                    if (getResources().getBoolean(R.bool.isTabletAndLandscape)) {
                        searchBarRoot.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                                ViewGroup.LayoutParams.WRAP_CONTENT));
                    }
                    if (webViewFragment != null) {
                        fragmentManager.beginTransaction()
                                .hide(webViewFragment)
                                .commit();
                    }
                } else {
                    duckLogo.setVisibility(View.VISIBLE);
                    if (getResources().getBoolean(R.bool.isTabletAndLandscape)) {
                        searchBarRoot.setLayoutParams(new LinearLayout.LayoutParams(550, ViewGroup.LayoutParams.WRAP_CONTENT));
                    }
                    if (webViewFragment != null) {
                        fragmentManager.beginTransaction()
                                .show(webViewFragment)
                                .commit();
                    }
                }
            }
        });

        eraseTextButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                searchBar.setText("");
                searchBar.requestFocus();
                manager.showSoftInput(searchBar, 0);
            }
        });

        if (darkTheme) {
            findViewById(R.id.search_bar_root).setBackgroundColor(
                    getResources().getColor(R.color.darkThemeColorPrimary));
            searchBarRoot.setBackground(getResources().getDrawable(R.drawable.search_field_bg_dark));
        }
    }

    @Override
    public void onSearchTermChange(String searchTerm) {
        searchBar.setText(searchTerm);
    }

    @Override
    public void onBackPressed() {
        if (searchBar.hasFocus() && webViewFragment != null) {
            webViewFragment.requestFocusOnWebView();
            searchBar.setText(latestTerm);
        } else {
            super.onBackPressed();
        }
    }

    void adapterUpdate() {
        new Thread(new Runnable() {
            @Override
            public void run() {
                ArrayList<HistoryItem> historyArrayList = (ArrayList<HistoryItem>) historyDatabase.historyDao().getAllSearchHistory();
                Collections.reverse(historyArrayList);
                ArrayList<String> historyArrayListStrings = new ArrayList<>();
                for (HistoryItem item : historyArrayList) {
                    historyArrayListStrings.add(item.getSearchTerm());
                }
                String[] historyArray = Arrays.copyOf(historyArrayListStrings.toArray(), historyArrayList.size(), String[].class);
                adapter = new AutoCompleteAdapter(SearchActivity.this, R.layout.auto_complete_item
                        , historyArray, searchBar);
                activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        searchBar.setAdapter(adapter);
                    }
                });
            }
        }).start();
    }

    void search(String searchTerm) {
        progressBar.setVisibility(View.VISIBLE);
        webViewFragment = WebViewFragment.newInstance(searchTerm, PrefManager.isHistoryEnabled(this));
        latestTerm = searchTerm;
        if (fromIntent) {
            searchBar.setText(latestTerm);
        }
        fragmentManager.beginTransaction()
                .replace(R.id.frame_layout, webViewFragment)
                .commit();
        searchBar.clearFocus();
        searchBar.setSelection(0);
        manager.hideSoftInputFromWindow(searchBar.getWindowToken(), 0);
        getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_HIDDEN);
    }

    @Override
    public void onItemClickListener(String searchTerm) {
        search(searchTerm);
    }


    @Override
    public void onWebViewError(int errorCode) {
        if (errorCode == WebViewClient.ERROR_HOST_LOOKUP || errorCode == WebViewClient.ERROR_TIMEOUT || errorCode == WebViewClient.ERROR_CONNECT) {
            progressBar.setVisibility(View.GONE);
            fragmentManager.beginTransaction()
                    .replace(R.id.frame_layout, new ErrorFragment())
                    .commit();
        }
    }

    @Override
    public void onReloadButtonClick() {
        search(latestTerm);
    }

    @Override
    public void onPageFinish() {
        adapterUpdate();
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                progressBar.setVisibility(View.GONE);
            }
        });
    }
}

File: app/src/main/java/io/duckduckgosearch/app/SearchActivity.kt
package io.duckduckgosearch.app

import android.app.Activity
import android.content.Intent
import android.content.res.Configuration
import android.os.Build
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.View.OnFocusChangeListener
import android.view.ViewGroup
import android.view.WindowManager
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.*
import android.widget.TextView.OnEditorActionListener
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.os.bundleOf
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.commit
import io.duckduckgosearch.app.ErrorFragment.OnReloadButtonClick
import io.duckduckgosearch.app.WebViewFragment.*
import io.duckduckgosearch.app.WebViewFragment.Companion.newInstance
import java.util.*

class SearchActivity : AppCompatActivity(), OnSearchTermChange, AutoCompleteAdapter.OnItemClickListener, OnWebViewError, OnReloadButtonClick, OnPageFinish {

    private lateinit var searchBar: DuckAutoCompleteTextView
    private lateinit var fragmentManager: FragmentManager
    private lateinit var progressBar: ProgressBar
    private lateinit var activity: Activity
    private lateinit var duckLogo: ImageView
    private lateinit var eraseTextButton: ImageButton
    private lateinit var searchBarRoot: RelativeLayout
    private var latestTerm: String = ""
    private lateinit var intentSearchTerm: String
    private var webViewFragment: WebViewFragment? = null
    private var errorFragment: ErrorFragment? = null
    private var adapter: AutoCompleteAdapter? = null
    private lateinit var manager: InputMethodManager
    private var historyDatabase: HistoryDatabase? = null
    private var fromIntent = false

    override fun onCreate(savedInstanceState: Bundle?) {

        super.onCreate(savedInstanceState)
        AppCompatDelegate.setCompatVectorFromResourcesEnabled(true)
        AppCompatDelegate.setDefaultNightMode(PrefManager.getThemeInt(this))

        this.transparentStatusBar()
        // Enable WebView debugging from PC (Chrome)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            WebView.setWebContentsDebuggingEnabled(true)
        }
        setContentView(R.layout.activity_search)
        historyDatabase = HistoryDatabase.getHistoryDatabase(this)
        manager = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
        duckLogo = findViewById(R.id.duck_logo)
        fragmentManager = supportFragmentManager
        activity = this
        searchBarRoot = findViewById(R.id.search_bar_edittext_root)
        progressBar = findViewById(R.id.search_progress)
        progressBar.visibility = View.GONE
        searchBar = findViewById(R.id.search_bar_edittext)
        adapterUpdate()
        eraseTextButton = findViewById(R.id.erase_button)
        val bundle = intent.extras
        if (bundle != null && bundle.containsKey("search_term")) {
            intentSearchTerm = bundle.getString("search_term").toString()
            fromIntent = true
            search(intentSearchTerm)
        }
        if (savedInstanceState == null && !fromIntent) {
            searchBar.requestFocus()
            eraseTextButton.visibility = View.GONE
        } else {
            window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_HIDDEN)
            if (!searchBar.hasFocus()) {
                duckLogo.visibility = View.VISIBLE
            }
        }
        searchBar.setOnEditorActionListener(OnEditorActionListener { v, actionId, _ ->
            if (v.text.toString() != "" && actionId == EditorInfo.IME_ACTION_DONE) {
                search(v.text.toString())
                adapter?.notifyDataSetChanged()
                return@OnEditorActionListener true
            }
            false
        })
        searchBar.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {}
            override fun onTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {}
            override fun afterTextChanged(p0: Editable?) {
                if (p0.toString() == "") {
                    eraseTextButton.visibility = View.GONE
                } else if (eraseTextButton.visibility == View.GONE) {
                    eraseTextButton.visibility = View.VISIBLE
                }
            }
        })
        searchBar.onFocusChangeListener = OnFocusChangeListener { _, hasFocus ->
            if (hasFocus) {
                duckLogo.visibility = View.GONE
                if (resources.getBoolean(R.bool.isTabletAndLandscape)) {
                    searchBarRoot.layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                            ViewGroup.LayoutParams.WRAP_CONTENT)
                }
                webViewFragment?.let {
                    fragmentManager.commit {
                        hide(it)
                    }
                }
                errorFragment?.let {
                    fragmentManager.commit {
                        hide(it)
                    }
                }
            } else {
                duckLogo.visibility = View.VISIBLE
                if (resources.getBoolean(R.bool.isTabletAndLandscape)) {
                    searchBarRoot.layoutParams = LinearLayout.LayoutParams(550, ViewGroup.LayoutParams.WRAP_CONTENT)
                }
                webViewFragment?.let {
                    fragmentManager.commit {
                        show(it)
                    }
                }
                errorFragment?.let {
                    fragmentManager.commit {
                        show(it)
                    }
                }
            }
        }
        eraseTextButton.setOnClickListener {
            searchBar.setText("")
            searchBar.requestFocus()
            manager.showSoftInput(searchBar, 0)
        }
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        webViewFragment = fragmentManager.getFragment(savedInstanceState,WebViewFragment.TAG) as WebViewFragment?
        errorFragment = fragmentManager.getFragment(savedInstanceState,ErrorFragment.TAG) as ErrorFragment?
        latestTerm = savedInstanceState.getString("latest_term","")
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString("latest_term",latestTerm)
        webViewFragment?.let {
            if (it.isAdded) {
                fragmentManager.putFragment(outState, WebViewFragment.TAG, it)
            }
        }
        errorFragment?.let {
            if (it.isAdded) {
                fragmentManager.putFragment(outState, ErrorFragment.TAG, it)
            }
        }
    }

    override fun onSearchTermChange(searchTerm: String?) {
        searchBar.setText(searchTerm)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        if (webViewFragment == null) {
            finish()
            startActivity(intent)
        } else {
            val relaunchIntent = Intent(this, SearchActivity::class.java)
            relaunchIntent.putExtras(bundleOf("search_term" to latestTerm))
            finish()
            startActivity(relaunchIntent)
        }
    }

    override fun onBackPressed() {
        if (searchBar.hasFocus() && (webViewFragment != null || errorFragment != null)) {
            searchBar.clearFocus()
            webViewFragment?.requestFocusOnWebView()
            searchBar.setText(latestTerm)
        } else {
            super.onBackPressed()
        }
    }

    private fun adapterUpdate() {
        Thread {
            val historyArrayList = historyDatabase?.historyDao()?.allSearchHistory as ArrayList<HistoryItem>
            historyArrayList.reverse()
            val historyArrayListStrings = ArrayList<String>()
            for (item in historyArrayList) {
                historyArrayListStrings.add(item.searchTerm)
            }
            val historyArray = Arrays.copyOf<String, Any>(historyArrayListStrings.toTypedArray(), historyArrayList.size, Array<String>::class.java)
            adapter = AutoCompleteAdapter(this@SearchActivity, R.layout.auto_complete_item, historyArray, searchBar)
            activity.runOnUiThread {
                searchBar.setAdapter(adapter)
                (searchBar.adapter as AutoCompleteAdapter).filter.filter("")
            }
        }.start()
    }

    private fun search(searchTerm: String) {
        progressBar.visibility = View.VISIBLE
        webViewFragment = newInstance(searchTerm, PrefManager.isHistoryEnabled(this))
        latestTerm = searchTerm
        if (fromIntent) {
            searchBar.setText(latestTerm)
        }
        webViewFragment?.let {
            fragmentManager.commit {
                replace(R.id.frame_layout, it)
            }
        }
        errorFragment = null
        searchBar.clearFocus()
        searchBar.setSelection(0)
        manager.hideSoftInputFromWindow(searchBar.windowToken, 0)
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_HIDDEN)
    }

    override fun onItemClickListener(searchTerm: String) {
        search(searchTerm)
    }

    override fun onWebViewError(errorCode: Int) {
        if (errorCode == WebViewClient.ERROR_HOST_LOOKUP || errorCode == WebViewClient.ERROR_TIMEOUT || errorCode == WebViewClient.ERROR_CONNECT) {
            progressBar.visibility = View.GONE
            errorFragment = ErrorFragment()
            fragmentManager.commit {
                replace(R.id.frame_layout, errorFragment!!)
            }
        }
    }

    override fun onReloadButtonClick() {
        if (latestTerm == "") {
            latestTerm = searchBar.text.toString()
        }
        search(latestTerm)
    }

    override fun onPageFinish() {
        adapterUpdate()
        runOnUiThread { progressBar.visibility = View.GONE }
    }
}